microsoft/PowerToys
# [FancyZones] Rework grid editor

## Summary of the Pull Request

This should fix many small issues with the grid layout editor. For the full list, see below.
The reason why this was all done in a single, large PR is that this work required rewriting a large part
of the backend code which maintains the grid layout, so it's hard to do incrementally.

**What is include in the PR:** 

* Do not allow new zone layouts to have zones smaller than a predefined amount - enough to display the zone index properly in the editor. Existing zone layouts which do not satisfy this requirement can still be edited, but there could be some visual glitches similar to the old editor.

![minsize](https://user-images.githubusercontent.com/9358361/110372116-316d8e80-804e-11eb-9438-6a5af6f5d795.gif)

* Resizers will now glide seamlessly across each other, and can also overlap.

**Before:**
![resizercrash](https://user-images.githubusercontent.com/20975774/107098551-c5f68e00-680f-11eb-9a0f-9ad0688f93db.gif)

**After:**
![gliding](https://user-images.githubusercontent.com/9358361/110371318-1b130300-804d-11eb-87fb-2b109ee6e381.gif)

* Don't allow splitting a zone into smaller zones if one of the resulting zones are too small.
* Show a visual cue (gray splitter) when the above would happen.

![nosplit](https://user-images.githubusercontent.com/9358361/110371486-56adcd00-804d-11eb-98fd-869a4fbfb717.gif)

* Splitters will now snap smoothly to zone borders, using the same magnetic snapping present in the canvas editor.

![smoothsnap](https://user-images.githubusercontent.com/9358361/110371669-9674b480-804d-11eb-80ce-eb7ed8b9a51a.gif)

* A new data representation simplifies code drastically.
* Now, the zones always logically occupy the entire screen area, and the spacing is added only when the zones are rendered. This is what allows for overlapping borders.
* FancyZonesLib code is tweaked to correspond to this new representation. There could be a small impact on how existing layouts are drawn, perhaps by a couple of pixels.

**How does someone test / validate:** 

Check all functionality of the grid editor (splitting zones, merging zones, resizing, adding many zones, canceling the edit).

For more stress testing:
Create a grid layout using the old version, create many small zones and set zone spacing to a large value, then open the layout in the new editor and try editing it. There could be some glitches but there should be no crashes.

Check whether the documentation for the editor should be updated.

## Quality Checklist

- [x] **Linked issue:** #9540 #9419 #5027

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.
